fenixedu-ist-giaf-invoices/src/main/java/pt/ist/fenixedu/giaf/invoices/ClientMap.java

```diff
@@ -0,0 +1,203 @@
+
+    private static final String MAP_FILENAME = "/afs/ist.utl.pt/ciist/fenix/fenix015/ist/giafVATNumberToClientIdMap.csv";
+
+    public ClientMap() {
+        try {
+            for (final String line : Files.readAllLines(new File(MAP_FILENAME).toPath())) {
+                final String[] parts = line.split("\t");
+
+                final String uVATNumber = parts[0];
+                final String clientId = parts[1];
+
+                put(uVATNumber, clientId);
+            }
+        } catch (final IOException e) {
+            throw new Error(e);
+        }
+    }
+
+    public void register(final String uVATNumber, final String clientId) {
+        put(uVATNumber, clientId);
+
+        try (final FileOutputStream stream = new FileOutputStream(MAP_FILENAME, true)) {
+            stream.write(uVATNumber.getBytes());
+            stream.write("\t".getBytes());
+            stream.write(clientId.getBytes());
+            stream.write("\n".getBytes());
+        } catch (final IOException e) {
+            throw new Error(e);
+        }
+    }
+
+    public String getClientId(final Person p) {
+        final String uVATNumber = uVATNumberFor(p);
+        return uVATNumber == null ? null : get(uVATNumber);
+    }
+
+    public boolean containsClient(final Person p) {
+        return getClientId(p) != null;
+    }
+
+    public static String uVATNumberFor(final Person person) {
+        final Country country = person.getCountry();
+        final String ssn = person.getSocialSecurityNumber();
+        final String vat = toVatNumber(ssn);
+        if (vat != null && isVatValidForPT(vat)) {
+            return "PT" + vat;
+        }
+        if (country == null || "PT".equals(country.getCode())) {
+            return null;
+        }
+        if (vat != null) {
+            return country.getCode() + vat;
+        }
+        final User user = person.getUser();
+        return user == null ? country.getCode() + makeUpSomeRandomNumber(person) : null;
+    }
+
+    private static String toVatNumber(final String ssn) {
+        return ssn == null ? null : ssn.startsWith("PT") ? ssn.substring(2) : ssn;
+    }
+
+    public static boolean isVatValidForPT(final String vat) {
+        if (vat.length() != 9) {
+            return false;
+        }
+        for (int i = 0; i < 9; i++) {
+            if (!Character.isDigit(vat.charAt(i))) {
+                return false;
+            }
+        }
+        if (Integer.parseInt(vat) <= 0) {
+            return false;
+        }
+        int sum = 0;
+        for (int i = 0; i < 8; i++) {
+            final int c = Character.getNumericValue(vat.charAt(i));
+            sum += c * (9 - i);
+        }
+        final int controleDigit = Character.getNumericValue(vat.charAt(8));
+        final int remainder = sum % 11;
+        int digit = 11 - remainder;
+        return digit > 9 ? controleDigit == 0 : digit == controleDigit;
+    }
+
+    private static String makeUpSomeRandomNumber(final Person person) {
```

Comment:
Not possible because we have a limited number of characters for the corresponding field in GIAF.